sql/core/src/test/scala/org/apache/spark/sql/DataFrameFunctionsSuite.scala

```diff
@@ -735,6 +735,44 @@ class DataFrameFunctionsSuite extends QueryTest with SharedSQLContext {
       df.selectExpr("array_contains(array(1, null), array(1, null)[0])"),
       Seq(Row(true), Row(true))
     )
+
+    checkAnswer(
+      df.selectExpr("array_contains(array(1), 1.23D)"),
+      Seq(Row(false), Row(false))
+    )
+
+    checkAnswer(
+      df.selectExpr("array_contains(array(1), 1.0D)"),
+      Seq(Row(true), Row(true))
+    )
+
+    checkAnswer(
+      df.selectExpr("array_contains(array(1.0D), 1)"),
+      Seq(Row(true), Row(true))
+    )
+
+    checkAnswer(
+      df.selectExpr("array_contains(array(1.23D), 1)"),
+      Seq(Row(false), Row(false))
+    )
+
+    checkAnswer(
+      df.selectExpr("array_contains(array(array(1)), array(1.0D))"),
+      Seq(Row(true), Row(true))
+    )
+
+    checkAnswer(
+      df.selectExpr("array_contains(array(array(1)), array(1.23D))"),
+      Seq(Row(false), Row(false))
+    )
+
+    intercept[AnalysisException] {
+      df.selectExpr("array_contains(array(1), 1.23)")
```

Comment:
Why does this fail?

Reply:
Also could you check the error message as well?

Reply:
@ueshin This fails since 1.23 is parsed as decimal(3, 2). And `findTightestCommonType` can not find a common type between integer and decimal(3,2).  I specifically added this case to draw it to your attention to see if we are okay with this behaviour :-). If we had it as `1.23D` then we would have been able to find a common type.

Reply:
@ueshin Did you want me to check the error message for all the cases ? or just a couple of them ?

Reply:
As for the behavior, we might be better to use `findWiderTypeWithoutStringPromotion` instead of `findTightestCommonType`.
And as for checking the error message, just those you added.

Reply:
@ueshin Thank you. I also thought of using `findWiderTypeWithoutStringPromotion` but later changed it to `findTightestCommonType`. One question i had was, can `findWiderTypeWithoutStringPromotion` do a lossy conversion. From the class description it seems only findTightestCommonType does a absolute safe casting ? Please let me know what you think ..

Reply:
@ueshin FYI - i just pushed the changes addressing your comments. I have kept `findTightestCommonType` as is for now. I will change it based on your input and adjust the tests accordingly.

Reply:
Good point. Yes, I can do a lossy conversion.
Seems like `BinaryComparison` uses wider DecimalType, so we could follow the behavior to use `findWiderTypeWithoutStringPromotion`.
cc @gatorsmile @cloud-fan

Reply:
Hmm, if we can `array_contains(array(1), '1')` in 2.3 as https://github.com/apache/spark/pull/22408#issuecomment-421223501, we should use `findWiderCommonType`?